tv_apps/tv-deck/js/channel_manager.js

```diff
@@ -25,7 +30,305 @@
+      this.isScanning = false;
+      this.playingTuner.sources = {};
+      if (sources.length === 0) {
+        console.log('Error, no source found!');
+        return;
+      }
+
+      sources.forEach(function initSource(source) {
+        this.playingTuner.sources[source.type] = {
+          source: source,
+          channels: [],
+          channelIndexHash: {}
+        };
+      }.bind(this));
+
+      if (!this.playingSourceType) {
+        this.playingSourceType = sources[0].type;
+      }
+
+      if (!this.playingSource) {
+        this.fire('error');
+        return;
+      }
+
+      this.scanChannels();
+    }.bind(this));
+  };
+
+  /**
+   * Retrieve all the currently available TV channels from current TV source.
+   * We have to scan the channels before calling getChannels.
+   */
+  proto.scanChannels = function cm_scanChannels() {
+    this.isScanning = true;
+    var source = this.playingSource.source;
+    this.playingSource.channels = [];
+    this.playingSource.channelIndexHash = {};
+    if (source.isScanning) {
+      return;
+    }
+
+    var onscanningstatechanged = function(event) {
+      var state = event.state;
+      switch (state) {
+        case 'completed':
+        /* falls through */
+        case 'stopped':
+          this._onScanningCompleted();
+          source.removeEventListener(
+                                'scanningstatechanged', onscanningstatechanged);
+          break;
+        default:
+          break;
+      }
+    }.bind(this);
+
+    source.addEventListener('scanningstatechanged', onscanningstatechanged);
+
+    source.startScanning({
+      isRescanned: true
+    });
+  };
+
+  /**
+   * Retrieve all the currently available TV channels from current TV source.
+   * Sort channels for channel switching.
+   */
+  proto._onScanningCompleted = function cm_onScanningCompleted() {
+    this.playingSource.source.getChannels().then(function onsuccess(channels) {
+      this.isScanning = false;
+      if (channels.length === 0) {
+        console.log('Error, no channel found!');
+        return;
+      }
+
+      // Sort channels. Channel number can be XX-XX-XX
+      channels.sort(function(channelA, channelB) {
+        return this._compareChannel(channelA.number, channelB.number);
+      }.bind(this));
+
+      var i;
+      for (i = 0; i < channels.length; i++) {
+        this.playingSource.channelIndexHash[channels[i].number] = i;
+        this.playingSource.channels[i] = {
+          channel: channels[i],
+          programs: {},
+        };
+      }
+
+      if (!this.playingChannelNumber) {
+        this.playingChannelNumber = channels[0].number;
+      }
+
+      if (!this.playingChannel) {
+        this.fire('error');
+        return;
+      }
+
+      this.fire('tune');
+    }.bind(this), function onerror(error) {
+      alert(error);
+    });
+  };
+
+  proto.setPlayingSource = function cm_setPlayingSource(callback) {
+    if (this.isScanning) {
```

Comment:
We should call the call back while we do early return. Otherwise, the user of this API will not know what's going on.